### ISSUE TITLE
hilo.b crashes on PowerPC; bad symbols

### ISSUE THREAD

**Issue by kernigh** [davidgiven/ack#72]
When I run `./hilo_b.osxppc`, I get an immediate segmentation fault.

```
$ ./hilo_b.osxppc
Segmentation fault
```

When I run `./hilo_b.linuxppc`, I get a segmentation fault after I input my name.

```
$ ./hilo_b.linuxppc 

Hi there! I'm written in B. Before we start, what is your name?
> George
Segmentation fault
```

For this bug report, I built the current master 03386b9 on my OpenBSD/amd64 machine, then copied `hilo_b.osxppc` and `hilo_b.linuxppc` to the machines that run them. I'm using my PowerBook G4 for OS X and qemu-system-ppc for Linux.

Some symbols have the wrong value. OS X `nm` reveals that `ioctl` points to the middle of a function in B's library:

```
$ nm -m hilo_b.osxppc | less
...
00001bd0 t __II8
00001cdc t __II9
00001d28 T _ioctl
00001e20 T _patch_addresses
00001e88 T _main
```

In gdb, `disas isatty` shows that isatty() calls ioctl() at 0x1d28; and `disas ioctl` shows the middle of a function. The program crashes because stdio calls isatty() to check if stdout is a tty. Then the program wrongly jumps to 0x1d28 and continues with a corrupt program counter until it crashes.

Linux `nm` shows that `brk` points outside the program text:

0000001b a EBADGTO
0000003f a EUNIMPL
00001b44 T _brk
10000078 T __m_a_i_n
100000c8 T _i_strcopy

The program doesn't seem to call brk(), so I don't know what causes the crash.

So far, I only have this problem with some B programs on PowerPC. I can reproduce this problem in OS X by simply calling putchar(). Here's the B program:

```
main()
{
	putchar('a');
	return (0);
}
```

Ran `ack -mosxppc -o simple simple.b`, copied it to my Mac, then `./simple` segfaulted, again because `ioctl` has the wrong value. This program still works in Linux.

I remember that B worked on PowerPC in the past. My recent changes to the assembler or linker might have broken something.

**Comment by davidgiven** [comment]
I've spent some time running tests, and it looks like the linuxppc tests have been failing for a while. I can't actually find a revision where they work.

Also, the tests are disabled on Travis --- enabling them causes lat/linux386/tests+newdispose_p to fail.

Building hilo.b for linuxppc and running it with qemu-ppc duplicates the crash, but disassembling the result with objdump shows ioctl at 10001d68, and it looks like a real function (starts with `mflr r0; addi r1, r1, -8`).

Could you attach a complete disassembly with objdump of hilo.b compiled for linuxppc? With command line. I'm curious to know how it differs from mine.

This is Linux on ARM --- I'll try again with amd64 later.

**Comment by kernigh** [comment]
These results are with master (still 03386b9); my only modifications are in Makefile, where I set `BUILDDIR = /home/kernigh/park/ack-build` and `CC = egcc`. I'm running OpenBSD 6.2 on amd64; egcc is from OpenBSD's package of gcc 4.9.4.

The linuxppc tests (from `_dummy_e_bin` to `pascalsets_p_bin`) still work for me. I don't have qemu-ppc, but I copy the tests from ack-build/obj/plat/linuxppc/tests/*/*bin to my Linux inside qemu-system-ppc and run them there. Each test prints `@@FINISHED` and exits 0.

This PowerPC Linux is Debian 8; its `objdump --version` starts with `GNU objdump (GNU Binutils for Debian) 2.25`.

- [hilo.txt](https://github.com/davidgiven/ack/files/1798956/hilo.txt) is the output of `objdump -D hilo_b.linuxppc`
- [hilo_b.linuxppc.gz](https://github.com/davidgiven/ack/files/1798959/hilo_b.linuxppc.gz) is the gzipped binary

**Comment by davidgiven** [comment]
Turns out that the underlying crash was an embarrassingly inept missing return value from B's `lchar()` --- that's now fixed.

The issues with the symbols are still weird. Your disassembly shows that ioctl points to the beginning of a function:

```
10001d9c <_ioctl>:
10001d9c:       7c 08 02 a6     mflr    r0
10001da0:       38 21 ff f8     addi    r1,r1,-8
10001da4:       90 41 00 00     stw     r2,0(r1)
10001da8:       90 01 00 04     stw     r0,4(r1)
10001dac:       38 41 00 00     addi    r2,r1,0
```
Can you double check that gcc's disas and objdump are showing the same values, for you? I wonder whether it's possible that our non-standard ELF generator is confusing gcc --- although that seems unlikely, as you're seeing this issue with Mach as well.

brk() does not appear in the disassembly at all, but it's in the same object file as sbrk(), so it should be there. (I don't believe led does per-function code removal.) If I add a junk function to sbrk.c, it appears with the correct address, so this is something special about brk() --- but brk() is referred to literally nowhere by any of the libraries in the linker command line...

**Comment by davidgiven** [comment]
No, wait, sbrk() calls it. In your disassembly (and I see this too), a call is generated where the code as appended to the end of read(). So the function is in the resulting file, but the symbol is wrong.

**Comment by davidgiven** [comment]
Fixed (in 4cb4bdc85f82080778b3d8a65e23ed1b81af930a).

It was nothing to do with B. Turns out that defining undefined globals with `-u` would add a global, but not increment the global count variable. changenames() uses this count to scan the globals and adjust the addresses, which means that for every symbol defined with `-u`, one symbol found via scanning files would be dropped off the end.

It just so happened that with hilo.b, _brk was the last symbol found via scanning.

This didn't affect actual code generation because (I think) the offsets here get added when the relocation is generated; the absolute value of the symbol is never used. So nobody ever noticed.

I think that this bug has been in the code since The Beginning (1985).

Thanks for spotting this!

**Comment by kernigh** [comment]
I didn't know that this was two different bugs (symbols and lchar). Thanks for fixing both. I confirm that `hilo_b.osxppc` now works on my Mac.
